token_metadata: keep the set of normal token owners as a member

bhalevy opened the pull request:
token_metadata: impl: keep the set of normal token owners as a member

We don't need to recalculate the unique set of normal token
everytime we change `_token_to_endpoint_map`.
Similarly, this doesn't have to be done in `get_all_endpoints`.

Instead we can maintain it inexpensively in
`remove_endpoint`, and let `count_normal_token_owners`
just return its size and `get_all_endpoints` just return
the saved set.

Closes #11128
Fixes #11146

bhalevy commented:
c80c532abd: rebased to pass CI

bhalevy commented:
given #11146, we should let update_normal_token yield

bhalevy commented:
@asias ping, please review

asias commented on locator/token_metadata.cc:
```diff
@@ -367,7 +366,8 @@ future<token_metadata_impl> token_metadata_impl::clone_only_token_map(bool clone
         }).then([this, &ret] {
-            ret._nr_normal_token_owners = _nr_normal_token_owners;
+            ret._normal_token_owners = _normal_token_owners;
+        }).then([this, &ret] {
```
Any reason to add the then here?

bhalevy replied:
The idea was to allow yielding, but it turns out we don't check `need_preempt` in .then when the future is ready.
so the reason is moot.

asias replied:
I see.

avikivity replied:
So why do it?

bhalevy replied:
In hindsight it's unneeded. I can remove it in v2.
(No CI ran apparently so pushing a new version should trigger one)

asias commented on locator/token_metadata.cc:
```diff
@@ -913,15 +913,16 @@ future<> token_metadata_impl::update_pending_ranges(
     std::unordered_set<inet_address> eps;
     for (auto [t, ep]: _token_to_endpoint_map) {
         eps.insert(ep);
+        co_await coroutine::maybe_yield();
```
This is going to make backport harder since we do not have coroutine::maybe_yield in 2021.1.

bhalevy replied:
2022.1 would be an easy target for backporting
but I'm not sure if the issue is severe enough to warrant backporting to 2021.1
We can deal with it if and when this is backported there.

See discussion in https://github.com/scylladb/scylla-enterprise/issues/2194#issuecomment-1202046266
There are quite a few optimizations around token_metadata (and particularly effective_replication_map) that aren't feasible for backporting there.

asias replied:
OK.

avikivity commented:
Let's have a new rule that says that if a patch fixes a non-bug issue ("X should do Y"), then we close it with Closes #vwxyz instead of Fixes #vwxyz, so there is less work to determine if it needs to be backported.

avikivity commented on locator/token_metadata.cc:
```diff
@@ -50,9 +50,9 @@ class token_metadata_impl final {
-    size_t _nr_normal_token_owners;
+
+    // Track the unique set of nodes in _token_to_endpoint_map
+    std::unordered_set<inet_address> _normal_token_owners;
```
Some things we can consider (not proposing for now):
 - switch to abseil btree containers for fewer allocations/branches
 - store all node vectors in a few larger vector and return std::span<>s to it
 - start storing node* instead of inet_address, pointing to a vector<node> i token_metadata

The last two require that we hold token_metadata_ptr in the read/write executors, but that's easy.

bhalevy replied:
I think that the node instances belong elsewhere, not in token_metadata, maybe in topology.
and then they can be reference by a shared ptr until they are removed from the cluster.

avikivity commented:
I don't see a CI run, what happened?

bhalevy commented:
> Let's have a new rule that says that if a patch fixes a non-bug issue ("X should do Y"), then we close it with Closes #vwxyz instead of Fixes #vwxyz, so there is less work to determine if it needs to be backported.

Makes sense

> I don't see a CI run, what happened?

Good question.  I'll push v2 (without the extraneous .then continuation) to trigger one.

bhalevy commented:
In v3 (0dfd92d0b3):
- rebased + retested
- updated git log message with "Closes" rather than "Fixes"
- deleted the extraneous `.then` continuation discussed in https://github.com/scylladb/scylladb/pull/11158#discussion_r935159907

avikivity commented:
Looks like CI is running.

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/1637/
